File: typescript/src/base/ExpressionParser.ts
import { Decimal } from 'decimal.js';
import { ValueRegistry } from './ValueRegistry';

export class Expression {
  static readonly AlwaysTrue = '1';
  static readonly AlwaysFalse = '0';

  static and(...expressions: string[]): string {
    if (expressions.length > 1) return `and(${expressions.join(', ')})`;
    if (expressions.length === 1) return expressions[0];
    return this.AlwaysTrue;
  }

  static or(...expressions: string[]): string {
    if (expressions.length > 1) return `or(${expressions.join(', ')})`;
    if (expressions.length === 1) return expressions[0];
    return this.AlwaysFalse;
  }
}

export interface ParseResult {
  function: (valueRegistry: ValueRegistry) => Decimal;
  dependencies: Set<string>;
}

class Node {
  value: string = '';
  left?: Node;
  right?: Node;
  parameters?: Node[]; // For function parameters
}

export class ExpressionParser {
  static parse(input: string): (valueRegistry: ValueRegistry) => Decimal {
    console.log(`Parsing expression: ${input}`);
    const tokens = this.tokenize(input);
    const ast = this.buildAST(tokens);

    return (valueRegistry: ValueRegistry) => {
      return this.evaluateAST(ast, valueRegistry);
    };
  }

  static parseWithDependencies(input: string): ParseResult {
    console.log(`Parsing expression: ${input}`);
    const tokens = this.tokenize(input);
    const ast = this.buildAST(tokens);
    const dependencies = new Set<string>();

    // Collect dependencies
    this.collectDependencies(ast, dependencies);

    return {
      function: (valueRegistry: ValueRegistry) => this.evaluateAST(ast, valueRegistry),
      dependencies,
    };
  }

  private static tokenize(input: string): string[] {
    const tokens: string[] = [];
    let currentToken = '';

    for (let i = 0; i < input.length; i++) {
      const c = input[i];

      // Whitespace terminates current token and is skipped
      if (/\s/.test(c)) {
        if (currentToken !== '') {
          tokens.push(currentToken);
          currentToken = '';
        }
        continue;
      }

      // Special characters are treated as independent tokens
      if (c === '(' || c === ')' || c === ',') {
        if (currentToken !== '') {
          tokens.push(currentToken);
          currentToken = '';
        }
        tokens.push(c);
        continue;
      }

      // Check for two-character operators
      if (i < input.length - 1) {
        const nextC = input[i + 1];
        const twoChars = c + nextC;
        if (twoChars === '>=' || twoChars === '<=' || twoChars === '==' || twoChars === '!=') {
          if (currentToken !== '') {
            tokens.push(currentToken);
            currentToken = '';
          }
          tokens.push(twoChars);
          i++; // Skip next character as it's already processed
          continue;
        }
      }

      // Check for single-character operators
      if (['+', '-', '*', '/', '^', '>', '<'].includes(c)) {
        if (currentToken !== '') {
          tokens.push(currentToken);
          currentToken = '';
        }
        tokens.push(c);
        continue;
      }

      // Add other characters to current token
      currentToken += c;
    }

    // Add final token if any
    if (currentToken !== '') {
      tokens.push(currentToken);
    }

    return tokens;
  }

  private static buildAST(tokens: string[]): Node {
    const stack: Node[] = [];
    const operators: string[] = [];
    const functionParams: number[] = []; // Track function parameter count

    for (let i = 0; i < tokens.length; i++) {
      const token = tokens[i];

      if (this.isFunction(token)) {
        operators.push(token);
        functionParams.push(0); // Initialize parameter count
      } else if (token === '(') {
        operators.push(token);
      } else if (token === ')') {
        while (operators.length > 0 && operators[operators.length - 1] !== '(') {
          this.processOperator(stack, operators.pop()!);
        }
        operators.pop(); // Remove '('

        if (operators.length > 0 && this.isFunction(operators[operators.length - 1])) {
          const func = operators.pop()!;
          const paramCount = functionParams.pop()! + 1;
          this.processFunction(stack, func, paramCount);
        }
      } else if (token === ',') {
        while (operators.length > 0 && operators[operators.length - 1] !== '(') {
          this.processOperator(stack, operators.pop()!);
        }
        if (functionParams.length > 0) {
          functionParams[functionParams.length - 1]++;
        }
      } else if (this.isOperator(token)) {
        while (
          operators.length > 0 &&
          operators[operators.length - 1] !== '(' &&
          this.getPrecedence(operators[operators.length - 1]) >= this.getPrecedence(token)
        ) {
          this.processOperator(stack, operators.pop()!);
        }
        operators.push(token);
      } else {
        stack.push({ value: token });
      }
    }

    while (operators.length > 0) {
      this.processOperator(stack, operators.pop()!);
    }

    return stack[0];
  }

  private static processOperator(stack: Node[], op: string): void {
    const right = stack.pop()!;
    const left = stack.pop()!;
    stack.push({
      value: op,
      left,
      right,
    });
  }

  private static processFunction(stack: Node[], func: string, paramCount: number): void {
    const parameters: Node[] = new Array(paramCount);
    for (let i = paramCount - 1; i >= 0; i--) {
      parameters[i] = stack.pop()!;
    }
    stack.push({ value: func, parameters });
  }

  private static isOperator(token: string): boolean {
    return ['+', '-', '*', '/', '^', '>', '<', '>=', '<=', '==', '!='].includes(token);
  }

  private static isFunction(token: string): boolean {
    return ['min', 'max', 'and', 'or'].includes(token);
  }

  private static getPrecedence(op: string): number {
    switch (op) {
      case '^':
        return 4;
      case '*':
      case '/':
        return 3;
      case '+':
      case '-':
        return 2;
      case '>':
      case '<':
      case '>=':
      case '<=':
      case '==':
      case '!=':
        return 1;
      default:
        return 0;
    }
  }

  private static evaluateAST(node: Node, valueRegistry: ValueRegistry): Decimal {
    if (!node) return new Decimal(0);

    // Handle functions
    if (node.parameters) {
      const params = node.parameters.map(p => this.evaluateAST(p, valueRegistry));
      switch (node.value) {
        case 'min':
          return Decimal.min(...params);
        case 'max':
          return Decimal.max(...params);
        case 'and':
          return params.every(p => p.greaterThanOrEqualTo(1)) ? new Decimal(1) : new Decimal(0);
        case 'or':
          return params.some(p => p.greaterThanOrEqualTo(1)) ? new Decimal(1) : new Decimal(0);
        default:
          throw new Error(`Unknown function: ${node.value}`);
      }
    }

    // Handle operators
    if (node.left && node.right) {
      const left = this.evaluateAST(node.left, valueRegistry);
      const right = this.evaluateAST(node.right, valueRegistry);

      switch (node.value) {
        case '+':
          return left.plus(right);
        case '-':
          return left.minus(right);
        case '*':
          return left.times(right);
        case '/':
          return left.dividedBy(right);
        case '^':
          return left.pow(right);
        case '>':
          return left.greaterThan(right) ? new Decimal(1) : new Decimal(0);
        case '<':
          return left.lessThan(right) ? new Decimal(1) : new Decimal(0);
        case '>=':
          return left.greaterThanOrEqualTo(right) ? new Decimal(1) : new Decimal(0);
        case '<=':
          return left.lessThanOrEqualTo(right) ? new Decimal(1) : new Decimal(0);
        case '==':
          return left.equals(right) ? new Decimal(1) : new Decimal(0);
        case '!=':
          return !left.equals(right) ? new Decimal(1) : new Decimal(0);
        default:
          throw new Error(`Unknown operator: ${node.value}`);
      }
    }

    // Handle literals and variables
    if (!isNaN(Number(node.value))) {
      return new Decimal(node.value);
    }

    return valueRegistry.getWrappedValue(node.value);
  }

  private static collectDependencies(node: Node, dependencies: Set<string>): void {
    if (!node) return;

    // Handle functions
    if (node.parameters) {
      node.parameters.forEach(p => this.collectDependencies(p, dependencies));
      return;
    }

    // Handle operators
    if (node.left && node.right) {
      this.collectDependencies(node.left, dependencies);
      this.collectDependencies(node.right, dependencies);
      return;
    }

    // Handle variables (non-numeric values)
    if (isNaN(Number(node.value))) {
      dependencies.add(node.value);
    }
  }
} 

File: BaseSystem/ExpressionParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BreakInfinity;

namespace BaseSystem
{
    public static class Expression
    {
        public static string AlwaysTrue = "1";
        public static string AlwaysFalse = "0";

        public static string And(params string[] expressions) =>
            expressions.Length > 1 ? $"and({string.Join(", ", expressions)})"
            : expressions.Length == 1 ? expressions[0]
            : AlwaysTrue;

        public static string Or(params string[] expressions) =>
            expressions.Length > 1 ? $"or({string.Join(", ", expressions)})"
            : expressions.Length == 1 ? expressions[0]
            : AlwaysFalse;
    }

    public struct ParseResult
    {
        public Func<ValueRegistry, BigDouble> Function { get; set; }
        public HashSet<string> Dependencies { get; set; }
    }

    public class ExpressionParser
    {
        // Input: 2 + level * 5 - 2^3
        public static Func<ValueRegistry, BigDouble> Parse(string input)
        {
            Console.WriteLine($"Parsing expression: {input}");
            var tokens = Tokenize(input);
            var ast = BuildAST(tokens);

            return (valueRegistry) =>
            {
                return EvaluateAST(ast, valueRegistry);
            };
        }

        public static ParseResult ParseWithDependencies(string input)
        {
            Console.WriteLine($"Parsing expression: {input}");
            var tokens = Tokenize(input);
            var ast = BuildAST(tokens);
            var dependencies = new HashSet<string>();

            // 의존성 수집
            CollectDependencies(ast, dependencies);

            return new ParseResult
            {
                Function = (valueRegistry) => EvaluateAST(ast, valueRegistry),
                Dependencies = dependencies,
            };
        }

        private static string[] Tokenize(string input)
        {
            var tokens = new List<string>();
            var currentToken = "";

            for (int i = 0; i < input.Length; i++)
            {
                var c = input[i];

                // 공백은 현재 토큰을 종료하고 건너뜁니다
                if (char.IsWhiteSpace(c))
                {
                    if (currentToken != "")
                    {
                        tokens.Add(currentToken);
                        currentToken = "";
                    }
                    continue;
                }

                // 특수 문자들은 독립적인 토큰으로 처리
                if (c is '(' or ')' or ',')
                {
                    if (currentToken != "")
                    {
                        tokens.Add(currentToken);
                        currentToken = "";
                    }
                    tokens.Add(c.ToString());
                    continue;
                }

                // 두 글자 연산자 체크
                if (i < input.Length - 1)
                {
                    var nextC = input[i + 1];
                    var twoChars = c.ToString() + nextC;
                    if (twoChars is ">=" or "<=" or "==" or "!=")
                    {
                        if (currentToken != "")
                        {
                            tokens.Add(currentToken);
                            currentToken = "";
                        }
                        tokens.Add(twoChars);
                        i++; // 다음 문자는 이미 처리했으므로 건너뜁니다
                        continue;
                    }
                }

                // 한 글자 연산자 체크
                if (c is '+' or '-' or '*' or '/' or '^' or '>' or '<')
                {
                    if (currentToken != "")
                    {
                        tokens.Add(currentToken);
                        currentToken = "";
                    }
                    tokens.Add(c.ToString());
                    continue;
                }

                // 나머지 문자들은 현재 토큰에 추가
                currentToken += c;
            }

            // 마지막 토큰이 남아있다면 추가
            if (currentToken != "")
            {
                tokens.Add(currentToken);
            }

            return tokens.ToArray();
        }

        private static Node BuildAST(string[] tokens)
        {
            var stack = new Stack<Node>();
            var operators = new Stack<string>();
            var functionParams = new Stack<int>(); // 함수 파라미터 개수 추적

            for (int i = 0; i < tokens.Length; i++)
            {
                var token = tokens[i];

                if (IsFunction(token))
                {
                    operators.Push(token);
                    functionParams.Push(0); // 파라미터 카운트 초기화
                }
                else if (token == "(")
                {
                    operators.Push(token);
                }
                else if (token == ")")
                {
                    while (operators.Count > 0 && operators.Peek() != "(")
                    {
                        ProcessOperator(stack, operators.Pop());
                    }
                    operators.Pop(); // "(" 제거

                    if (operators.Count > 0 && IsFunction(operators.Peek()))
                    {
                        var func = operators.Pop();
                        var paramCount = functionParams.Pop() + 1;
                        ProcessFunction(stack, func, paramCount);
                    }
                }
                else if (token == ",")
                {
                    while (operators.Count > 0 && operators.Peek() != "(")
                    {
                        ProcessOperator(stack, operators.Pop());
                    }
                    if (functionParams.Count > 0)
                    {
                        functionParams.Push(functionParams.Pop() + 1);
                    }
                }
                else if (IsOperator(token))
                {
                    while (
                        operators.Count > 0
                        && operators.Peek() != "("
                        && GetPrecedence(operators.Peek()) >= GetPrecedence(token)
                    )
                    {
                        ProcessOperator(stack, operators.Pop());
                    }
                    operators.Push(token);
                }
                else
                {
                    stack.Push(new Node { Value = token });
                }
            }

            while (operators.Count > 0)
            {
                ProcessOperator(stack, operators.Pop());
            }

            return stack.Pop();
        }

        private static void ProcessOperator(Stack<Node> stack, string op)
        {
            var right = stack.Pop();
            var left = stack.Pop();
            stack.Push(
                new Node
                {
                    Value = op,
                    Left = left,
                    Right = right,
                }
            );
        }

        private static void ProcessFunction(Stack<Node> stack, string func, int paramCount)
        {
            var parameters = new Node[paramCount];
            for (int i = paramCount - 1; i >= 0; i--)
            {
                parameters[i] = stack.Pop();
            }
            stack.Push(new Node { Value = func, Parameters = parameters });
        }

        private static bool IsOperator(string token)
        {
            return token
                is "+"
                    or "-"
                    or "*"
                    or "/"
                    or "^"
                    or ">"
                    or "<"
                    or ">="
                    or "<="
                    or "=="
                    or "!=";
        }

        private static bool IsFunction(string token)
        {
            return token is "min" or "max" or "and" or "or";
        }

        private static int GetPrecedence(string op) =>
            op switch
            {
                "==" or "!=" => 1,
                ">" or "<" or ">=" or "<=" => 2,
                "+" or "-" => 3,
                "*" or "/" => 4,
                "^" => 5,
                _ => 0,
            };

        private static BigDouble EvaluateAST(Node node, ValueRegistry valueRegistry)
        {
            if (node == null)
                return 0;

            // 숫자인 경우
            if (double.TryParse(node.Value, out double number))
                return number;

            // 연산자나 함수가 아닌 문자열인 경우 valueRegistry에서 값을 가져옴
            if (!IsOperator(node.Value) && !IsFunction(node.Value) && node.Parameters == null)
                return valueRegistry.Get(node.Value).GetValue(valueRegistry);

            if (node.Parameters != null)
            {
                // Use lazy evaluation to avoid unnecessary computations
                return node.Value switch
                {
                    "min" => node.Parameters.Select(p => EvaluateAST(p, valueRegistry)).Min(),
                    "max" => node.Parameters.Select(p => EvaluateAST(p, valueRegistry)).Max(),
                    "and" => node.Parameters.Any(p => EvaluateAST(p, valueRegistry) < 1) ? 0 : 1,
                    "or" => node.Parameters.Any(p => EvaluateAST(p, valueRegistry) >= 1) ? 1 : 0,
                    _ => throw new ArgumentException($"Unknown function: {node.Value}"),
                };
            }

            return node.Value switch
            {
                "+" => EvaluateAST(node.Left, valueRegistry)
                    + EvaluateAST(node.Right, valueRegistry),
                "-" => EvaluateAST(node.Left, valueRegistry)
                    - EvaluateAST(node.Right, valueRegistry),
                "*" => EvaluateAST(node.Left, valueRegistry)
                    * EvaluateAST(node.Right, valueRegistry),
                "/" => EvaluateAST(node.Left, valueRegistry)
                    / EvaluateAST(node.Right, valueRegistry),
                "^" => BigDouble.Pow(
                    EvaluateAST(node.Left, valueRegistry),
                    EvaluateAST(node.Right, valueRegistry)
                ),
                ">" => EvaluateAST(node.Left, valueRegistry)
                > EvaluateAST(node.Right, valueRegistry)
                    ? 1
                    : 0,
                "<" => EvaluateAST(node.Left, valueRegistry)
                < EvaluateAST(node.Right, valueRegistry)
                    ? 1
                    : 0,
                ">=" => EvaluateAST(node.Left, valueRegistry)
                >= EvaluateAST(node.Right, valueRegistry)
                    ? 1
                    : 0,
                "<=" => EvaluateAST(node.Left, valueRegistry)
                <= EvaluateAST(node.Right, valueRegistry)
                    ? 1
                    : 0,
                "==" => EvaluateAST(node.Left, valueRegistry)
                == EvaluateAST(node.Right, valueRegistry)
                    ? 1
                    : 0,
                "!=" => EvaluateAST(node.Left, valueRegistry)
                != EvaluateAST(node.Right, valueRegistry)
                    ? 1
                    : 0,
                _ => throw new ArgumentException($"Unknown operator: {node.Value}"),
            };
        }

        private static void CollectDependencies(Node node, HashSet<string> dependencies)
        {
            if (node == null)
                return;

            // 숫자가 아니고, 연산자나 함수도 아닌 경우 의존성으로 추가
            if (
                !double.TryParse(node.Value, out _)
                && !IsOperator(node.Value)
                && !IsFunction(node.Value)
            )
            {
                dependencies.Add(node.Value);
            }

            // 재귀적으로 자식 노드들의 의존성도 수집
            if (node.Parameters != null)
            {
                foreach (var param in node.Parameters)
                {
                    CollectDependencies(param, dependencies);
                }
            }
            else
            {
                CollectDependencies(node.Left, dependencies);
                CollectDependencies(node.Right, dependencies);
            }
        }

        private class Node
        {
            public string Value { get; set; }
            public Node Left { get; set; }
            public Node Right { get; set; }
            public Node[] Parameters { get; set; } // 함수 파라미터용
        }
    }
}
